cdap-notifications/src/main/java/co/cask/cdap/notifications/service/inmemory/InMemoryNotificationService.java

```diff
@@ -0,0 +1,188 @@
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.gson.Gson;
+import com.google.gson.JsonElement;
+import com.google.inject.Inject;
+import org.apache.twill.common.Cancellable;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Type;
+import java.util.Collection;
+import java.util.concurrent.Executor;
+
+/**
+ * In-memory Notification service that pushes notifications to subscribers.
+ */
+public class InMemoryNotificationService extends AbstractIdleService implements NotificationService {
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryNotificationService.class);
+  private static final Gson GSON = new Gson();
+
+  private final Multimap<NotificationFeed, NotificationCaller<?>> subscribers;
+
+  private final DatasetFramework dsFramework;
+  private final TransactionSystemClient transactionSystemClient;
+  private final NotificationFeedManager feedManager;
+
+  @Inject
+  public InMemoryNotificationService(DatasetFramework dsFramework, TransactionSystemClient transactionSystemClient,
+                                     NotificationFeedManager feedManager) {
+    this.dsFramework = dsFramework;
+    this.transactionSystemClient = transactionSystemClient;
+    this.feedManager = feedManager;
+    this.subscribers = Multimaps.synchronizedMultimap(HashMultimap.<NotificationFeed, NotificationCaller<?>>create());
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    // No-op
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    // No-op
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification)
+    throws NotificationException {
+    return publish(feed, notification, notification.getClass());
+  }
+
+  @Override
+  public <N> ListenableFuture<N> publish(NotificationFeed feed, N notification, Type notificationType)
+    throws NotificationException {
+    Collection<NotificationCaller<?>> callers = subscribers.get(feed);
+    synchronized (subscribers) {
+      callers = ImmutableList.copyOf(callers);
+    }
+    for (NotificationCaller caller : callers) {
+      caller.received(notification, new BasicNotificationContext(dsFramework, transactionSystemClient));
+    }
+    return Futures.immediateFuture(notification);
+  }
+
+  /**
+   * Publishes a notification as a {@link JsonElement} to a {@code feed}.
+   *
+   * @param feed feed to publish notification to
+   * @param notificationJson json element representing the notification to publish
+   * @return a future describing the state of publishing this notification to all subscribers.
+   * @see #publish(NotificationFeed, Object, Type)
+   */
+  public ListenableFuture<JsonElement> publish(NotificationFeed feed, JsonElement notificationJson) {
```

Comment:
Rather than exposing this as a public method, it's better to have a abstract base class for both in-memory and Kafka implementation to inherit from. The base class would have the `subscribe` methods implemented. Something like:

```java
abstract class AbstractNotificationService extends AbstractIdleService implements NotificationService {
   protected void notificationReceived(NotificationFeed feed, JsonElement notificationJson) {
     // Similar what you have in this publish method.
   }

   public <N> Cancellable subscribe(NotificationFeed feed, NotificationHandler<N> handler, Executor executor) {
      // Just like the logic in this class
   }
}
```

Then in the in-memory implementation would call the `notificationReceived` method on publish (maybe from an internal executor to detach it from the caller thread). In the Kafka case, it will be called from the Kafka messageCallback.